File: php/src/Resources/UserResource.php
<?php

declare(strict_types=1);

namespace JewelMusic\Resources;

/**
 * User resource for profile management, preferences, and account settings
 */
class UserResource extends BaseResource
{
    /**
     * Get user profile information
     *
     * @return array User profile data
     */
    public function getProfile(): array
    {
        $response = $this->httpClient->get('/user/profile');
        return $this->extractData($response);
    }

    /**
     * Update user profile information
     *
     * @param array $updates Profile updates
     *                      - name: User display name
     *                      - bio: User biography
     *                      - website: User website URL
     *                      - location: User location
     *                      - socialLinks: Social media links
     *                      - artistInfo: Artist-specific information
     * @return array Updated profile data
     */
    public function updateProfile(array $updates): array
    {
        $response = $this->httpClient->put('/user/profile', $this->filterNullValues($updates));
        return $this->extractData($response);
    }

    /**
     * Upload user avatar image
     *
     * @param string|resource $avatarFile Avatar image file
     * @param string $filename Avatar filename
     * @return array Avatar upload result
     */
    public function uploadAvatar($avatarFile, string $filename): array
    {
        $this->validateRequired([
            'avatarFile' => $avatarFile,
            'filename' => $filename
        ], ['avatarFile', 'filename']);
        
        $response = $this->httpClient->uploadFile('/user/avatar', $avatarFile, $filename);
        return $this->extractData($response);
    }

    /**
     * Get user preferences and settings
     *
     * @return array User preferences
     */
    public function getPreferences(): array
    {
        $response = $this->httpClient->get('/user/preferences');
        return $this->extractData($response);
    }

    /**
     * Update user preferences and settings
     *
     * @param array $preferences User preferences
     *                          - notifications: Notification preferences
     *                          - ui: UI preferences
     *                          - privacy: Privacy settings
     *                          - api: API preferences
     * @return array Updated preferences
     */
    public function updatePreferences(array $preferences): array
    {
        $response = $this->httpClient->put('/user/preferences', $this->filterNullValues($preferences));
        return $this->extractData($response);
    }

    /**
     * Get list of user's API keys
     *
     * @return array List of API keys
     */
    public function getApiKeys(): array
    {
        $response = $this->httpClient->get('/user/api-keys');
        return $this->extractData($response);
    }

    /**
     * Create a new API key
     *
     * @param string $name API key name
     * @param array $permissions API key permissions
     *                          - scopes: Permission scopes
     *                          - rateLimit: Rate limit for this key
     *                          - ipRestrictions: IP address restrictions
     *                          - expiresAt: Expiration date
     *                          - description: Key description
     * @return array Created API key data
     */
    public function createApiKey(string $name, array $permissions = []): array
    {
        $this->validateRequired(['name' => $name], ['name']);
        
        $data = array_merge(['name' => $name], $this->filterNullValues($permissions));
        $response = $this->httpClient->post('/user/api-keys', $data);
        return $this->extractData($response);
    }

    /**
     * Update an existing API key
     *
     * @param string $keyId API key ID
     * @param array $updates API key updates
     * @return array Updated API key data
     */
    public function updateApiKey(string $keyId, array $updates): array
    {
        $this->validateRequired(['keyId' => $keyId], ['keyId']);
        
        $response = $this->httpClient->put("/user/api-keys/{$keyId}", $this->filterNullValues($updates));
        return $this->extractData($response);
    }

    /**
     * Revoke (delete) an API key
     *
     * @param string $keyId API key ID
     * @return array Revocation confirmation
     */
    public function revokeApiKey(string $keyId): array
    {
        $this->validateRequired(['keyId' => $keyId], ['keyId']);
        
        $response = $this->httpClient->delete("/user/api-keys/{$keyId}");
        return $this->extractData($response);
    }

    /**
     * Get detailed API usage statistics
     *
     * @param array $options Usage statistics options
     *                      - period: Statistics period
     *                      - startDate: Start date for statistics
     *                      - endDate: End date for statistics
     *                      - groupBy: Data grouping method
     *                      - includeBreakdown: Include detailed breakdown
     *                      - apiKeyId: Specific API key to analyze
     * @return array Usage statistics
     */
    public function getUsageStats(array $options = []): array
    {
        $params = $this->buildParams([], $options);
        $response = $this->httpClient->get('/user/usage', $params);
        return $this->extractData($response);
    }

    /**
     * Get billing information and invoices
     *
     * @param array $options Billing query options
     *                      - includeInvoices: Include invoice history
     *                      - invoiceLimit: Number of invoices to return
     *                      - invoiceStatus: Filter by invoice status
     * @return array Billing information
     */
    public function getBilling(array $options = []): array
    {
        $params = $this->buildParams([], $options);
        $response = $this->httpClient->get('/user/billing', $params);
        return $this->extractData($response);
    }

    /**
     * Update billing information
     *
     * @param array $billingData Updated billing information
     *                          - paymentMethod: Payment method details
     *                          - billingAddress: Billing address
     *                          - taxId: Tax identification number
     *                          - company: Company information
     * @return array Updated billing information
     */
    public function updateBilling(array $billingData): array
    {
        $response = $this->httpClient->put('/user/billing', $this->filterNullValues($billingData));
        return $this->extractData($response);
    }

    /**
     * Download invoice by ID
     *
     * @param string $invoiceId Invoice ID
     * @param string $format Download format (pdf, json)
     * @return array Invoice download data
     */
    public function downloadInvoice(string $invoiceId, string $format = 'pdf'): array
    {
        $this->validateRequired(['invoiceId' => $invoiceId], ['invoiceId']);
        
        $params = ['format' => $format];
        $response = $this->httpClient->get("/user/billing/invoices/{$invoiceId}", $params);
        return $this->extractData($response);
    }

    /**
     * Get account limits and quotas
     *
     * @return array Account limits information
     */
    public function getLimits(): array
    {
        $response = $this->httpClient->get('/user/limits');
        return $this->extractData($response);
    }

    /**
     * Get user activity log
     *
     * @param array $filters Activity filters
     *                      - startDate: Start date for activity log
     *                      - endDate: End date for activity log
     *                      - actionType: Type of actions to include
     *                      - limit: Number of entries to return
     * @return array Activity log entries
     */
    public function getActivityLog(array $filters = []): array
    {
        $params = $this->buildParams([], $filters);
        $response = $this->httpClient->get('/user/activity', $params);
        return $this->extractData($response);
    }

    /**
     * Get user notifications
     *
     * @param array $filters Notification filters
     *                      - unreadOnly: Show only unread notifications
     *                      - type: Notification type filter
     *                      - limit: Number of notifications to return
     * @return array User notifications
     */
    public function getNotifications(array $filters = []): array
    {
        $params = $this->buildParams([], $filters);
        $response = $this->httpClient->get('/user/notifications', $params);
        return $this->extractData($response);
    }

    /**
     * Mark notifications as read
     *
     * @param array $notificationIds Notification IDs to mark as read
     * @return array Update confirmation
     */
    public function markNotificationsRead(array $notificationIds): array
    {
        $data = ['notificationIds' => $notificationIds];
        $response = $this->httpClient->post('/user/notifications/mark-read', $data);
        return $this->extractData($response);
    }

    /**
     * Update notification preferences
     *
     * @param array $preferences Notification preferences
     *                          - email: Email notification settings
     *                          - push: Push notification settings
     *                          - sms: SMS notification settings
     * @return array Updated notification preferences
     */
    public function updateNotificationPreferences(array $preferences): array
    {
        $response = $this->httpClient->put('/user/notifications/preferences', $this->filterNullValues($preferences));
        return $this->extractData($response);
    }

    /**
     * Delete user account
     *
     * @param string $confirmEmail Email confirmation for account deletion
     * @param string $reason Reason for account deletion
     * @param bool $deleteData Whether to delete all user data
     * @return array Deletion confirmation
     */
    public function deleteAccount(string $confirmEmail, string $reason = '', bool $deleteData = true): array
    {
        $this->validateRequired(['confirmEmail' => $confirmEmail], ['confirmEmail']);
        
        $data = [
            'confirmEmail' => $confirmEmail,
            'reason' => $reason,
            'deleteData' => $deleteData
        ];
        
        $response = $this->httpClient->delete('/user/account', $data);
        return $this->extractData($response);
    }

    /**
     * Export user data
     *
     * @param array $options Export options
     *                      - format: Export format (json, csv)
     *                      - includeMetadata: Include metadata in export
     *                      - includeTracks: Include track data
     *                      - includeAnalytics: Include analytics data
     *                      - email: Email address for delivery
     * @return array Export job data
     */
    public function exportData(array $options = []): array
    {
        $response = $this->httpClient->post('/user/export', $this->filterNullValues($options));
        return $this->extractData($response);
    }

    /**
     * Get user's subscription information
     *
     * @return array Subscription details
     */
    public function getSubscription(): array
    {
        $response = $this->httpClient->get('/user/subscription');
        return $this->extractData($response);
    }

    /**
     * Upgrade or change user subscription
     *
     * @param array $subscriptionData Subscription details
     *                               - plan: Subscription plan
     *                               - billing: Billing frequency
     * @return array Updated subscription information
     */
    public function updateSubscription(array $subscriptionData): array
    {
        $response = $this->httpClient->put('/user/subscription', $this->filterNullValues($subscriptionData));
        return $this->extractData($response);
    }

    /**
     * Cancel user subscription
     *
     * @param array $cancellationData Cancellation details
     *                               - reason: Cancellation reason
     *                               - feedback: User feedback
     *                               - effective: Effective cancellation date
     * @return array Cancellation confirmation
     */
    public function cancelSubscription(array $cancellationData = []): array
    {
        $response = $this->httpClient->post('/user/subscription/cancel', $this->filterNullValues($cancellationData));
        return $this->extractData($response);
    }
}

File: ruby/lib/jewelmusic/resources/user_resource.rb
# frozen_string_literal: true

module JewelMusic
  module Resources
    ##
    # User resource for profile management, preferences, and account settings
    #
    class UserResource < BaseResource
      ##
      # Get user profile information
      #
      # @return [Hash] User profile data
      #
      def get_profile
        response = http_client.get('/user/profile')
        extract_data(response)
      end

      ##
      # Update user profile information
      #
      # @param updates [Hash] Profile updates
      # @option updates [String] :name User display name
      # @option updates [String] :bio User biography
      # @option updates [String] :website User website URL
      # @option updates [String] :location User location
      # @option updates [Hash] :social_links Social media links
      # @option updates [Hash] :artist_info Artist-specific information
      # @return [Hash] Updated profile data
      #
      def update_profile(updates)
        response = http_client.put('/user/profile', filter_nil_values(updates))
        extract_data(response)
      end

      ##
      # Upload user avatar image
      #
      # @param avatar_file [String, File, IO] Avatar image file
      # @param filename [String] Avatar filename
      # @return [Hash] Avatar upload result
      #
      def upload_avatar(avatar_file, filename)
        validate_required({
          avatar_file: avatar_file,
          filename: filename
        }, [:avatar_file, :filename])
        
        prepared_file = prepare_file(avatar_file)
        response = http_client.upload_file('/user/avatar', prepared_file, filename)
        extract_data(response)
      end

      ##
      # Get user preferences and settings
      #
      # @return [Hash] User preferences
      #
      def get_preferences
        response = http_client.get('/user/preferences')
        extract_data(response)
      end

      ##
      # Update user preferences and settings
      #
      # @param preferences [Hash] User preferences
      # @option preferences [Hash] :notifications Notification preferences
      # @option preferences [Hash] :ui UI preferences
      # @option preferences [Hash] :privacy Privacy settings
      # @option preferences [Hash] :api API preferences
      # @return [Hash] Updated preferences
      #
      def update_preferences(preferences)
        response = http_client.put('/user/preferences', filter_nil_values(preferences))
        extract_data(response)
      end

      ##
      # Get list of user's API keys
      #
      # @return [Hash] List of API keys
      #
      def get_api_keys
        response = http_client.get('/user/api-keys')
        extract_data(response)
      end

      ##
      # Create a new API key
      #
      # @param name [String] API key name
      # @param permissions [Hash] API key permissions
      # @option permissions [Array<String>] :scopes Permission scopes
      # @option permissions [Integer] :rate_limit Rate limit for this key
      # @option permissions [Array<String>] :ip_restrictions IP address restrictions
      # @option permissions [String] :expires_at Expiration date
      # @option permissions [String] :description Key description
      # @return [Hash] Created API key data
      #
      def create_api_key(name, permissions = {})
        validate_required({ name: name }, [:name])
        
        data = { name: name }.merge(filter_nil_values(permissions))
        response = http_client.post('/user/api-keys', data)
        extract_data(response)
      end

      ##
      # Update an existing API key
      #
      # @param key_id [String] API key ID
      # @param updates [Hash] API key updates
      # @return [Hash] Updated API key data
      #
      def update_api_key(key_id, updates)
        validate_required({ key_id: key_id }, [:key_id])
        
        response = http_client.put("/user/api-keys/#{key_id}", filter_nil_values(updates))
        extract_data(response)
      end

      ##
      # Revoke (delete) an API key
      #
      # @param key_id [String] API key ID
      # @return [Hash] Revocation confirmation
      #
      def revoke_api_key(key_id)
        validate_required({ key_id: key_id }, [:key_id])
        
        response = http_client.delete("/user/api-keys/#{key_id}")
        extract_data(response)
      end

      ##
      # Get detailed API usage statistics
      #
      # @param options [Hash] Usage statistics options
      # @option options [String] :period Statistics period
      # @option options [String] :start_date Start date for statistics
      # @option options [String] :end_date End date for statistics
      # @option options [String] :group_by Data grouping method
      # @option options [Boolean] :include_breakdown Include detailed breakdown
      # @option options [String] :api_key_id Specific API key to analyze
      # @return [Hash] Usage statistics
      #
      def get_usage_stats(options = {})
        params = build_params({}, options)
        response = http_client.get('/user/usage', params)
        extract_data(response)
      end

      ##
      # Get billing information and invoices
      #
      # @param options [Hash] Billing query options
      # @option options [Boolean] :include_invoices Include invoice history
      # @option options [Integer] :invoice_limit Number of invoices to return
      # @option options [String] :invoice_status Filter by invoice status
      # @return [Hash] Billing information
      #
      def get_billing(options = {})
        params = build_params({}, options)
        response = http_client.get('/user/billing', params)
        extract_data(response)
      end

      ##
      # Update billing information
      #
      # @param billing_data [Hash] Updated billing information
      # @option billing_data [Hash] :payment_method Payment method details
      # @option billing_data [Hash] :billing_address Billing address
      # @option billing_data [String] :tax_id Tax identification number
      # @option billing_data [String] :company Company information
      # @return [Hash] Updated billing information
      #
      def update_billing(billing_data)
        response = http_client.put('/user/billing', filter_nil_values(billing_data))
        extract_data(response)
      end

      ##
      # Download invoice by ID
      #
      # @param invoice_id [String] Invoice ID
      # @param format [String] Download format (pdf, json)
      # @return [Hash] Invoice download data
      #
      def download_invoice(invoice_id, format = 'pdf')
        validate_required({ invoice_id: invoice_id }, [:invoice_id])
        
        params = { format: format }
        response = http_client.get("/user/billing/invoices/#{invoice_id}", params)
        extract_data(response)
      end

      ##
      # Get account limits and quotas
      #
      # @return [Hash] Account limits information
      #
      def get_limits
        response = http_client.get('/user/limits')
        extract_data(response)
      end

      ##
      # Get user activity log
      #
      # @param filters [Hash] Activity filters
      # @option filters [String] :start_date Start date for activity log
      # @option filters [String] :end_date End date for activity log
      # @option filters [String] :action_type Type of actions to include
      # @option filters [Integer] :limit Number of entries to return
      # @return [Hash] Activity log entries
      #
      def get_activity_log(filters = {})
        params = build_params({}, filters)
        response = http_client.get('/user/activity', params)
        extract_data(response)
      end

      ##
      # Get user notifications
      #
      # @param filters [Hash] Notification filters
      # @option filters [Boolean] :unread_only Show only unread notifications
      # @option filters [String] :type Notification type filter
      # @option filters [Integer] :limit Number of notifications to return
      # @return [Hash] User notifications
      #
      def get_notifications(filters = {})
        params = build_params({}, filters)
        response = http_client.get('/user/notifications', params)
        extract_data(response)
      end

      ##
      # Mark notifications as read
      #
      # @param notification_ids [Array<String>] Notification IDs to mark as read
      # @return [Hash] Update confirmation
      #
      def mark_notifications_read(notification_ids)
        validate_required({ notification_ids: notification_ids }, [:notification_ids])
        data = { notification_ids: notification_ids }
        response = http_client.post('/user/notifications/mark-read', data)
        extract_data(response)
      end

      ##
      # Update notification preferences
      #
      # @param preferences [Hash] Notification preferences
      # @option preferences [Hash] :email Email notification settings
      # @option preferences [Hash] :push Push notification settings
      # @option preferences [Hash] :sms SMS notification settings
      # @return [Hash] Updated notification preferences
      #
      def update_notification_preferences(preferences)
        response = http_client.put('/user/notifications/preferences', filter_nil_values(preferences))
        extract_data(response)
      end

      ##
      # Delete user account
      #
      # @param confirm_email [String] Email confirmation for account deletion
      # @param reason [String] Reason for account deletion
      # @param delete_data [Boolean] Whether to delete all user data
      # @return [Hash] Deletion confirmation
      #
      def delete_account(confirm_email, reason = '', delete_data = true)
        validate_required({ confirm_email: confirm_email }, [:confirm_email])
        
        data = {
          confirm_email: confirm_email,
          reason: reason,
          delete_data: delete_data
        }
        
        response = http_client.delete('/user/account', data)
        extract_data(response)
      end

      ##
      # Export user data
      #
      # @param options [Hash] Export options
      # @option options [String] :format Export format (json, csv)
      # @option options [Boolean] :include_metadata Include metadata in export
      # @option options [Boolean] :include_tracks Include track data
      # @option options [Boolean] :include_analytics Include analytics data
      # @option options [String] :email Email address for delivery
      # @return [Hash] Export job data
      #
      def export_data(options = {})
        response = http_client.post('/user/export', filter_nil_values(options))
        extract_data(response)
      end

      ##
      # Get user's subscription information
      #
      # @return [Hash] Subscription details
      #
      def get_subscription
        response = http_client.get('/user/subscription')
        extract_data(response)
      end

      ##
      # Upgrade or change user subscription
      #
      # @param subscription_data [Hash] Subscription details
      # @option subscription_data [String] :plan Subscription plan
      # @option subscription_data [String] :billing Billing frequency
      # @return [Hash] Updated subscription information
      #
      def update_subscription(subscription_data)
        response = http_client.put('/user/subscription', filter_nil_values(subscription_data))
        extract_data(response)
      end

      ##
      # Cancel user subscription
      #
      # @param cancellation_data [Hash] Cancellation details
      # @option cancellation_data [String] :reason Cancellation reason
      # @option cancellation_data [String] :feedback User feedback
      # @option cancellation_data [String] :effective Effective cancellation date
      # @return [Hash] Cancellation confirmation
      #
      def cancel_subscription(cancellation_data = {})
        response = http_client.post('/user/subscription/cancel', filter_nil_values(cancellation_data))
        extract_data(response)
      end
    end
  end
end